parsing/ast_helper.ml

```diff
@@ -242,6 +242,32 @@ module Mty = struct
   let with_ ?loc ?attrs a b = mk ?loc ?attrs (Pmty_with (a, b))
   let typeof_ ?loc ?attrs a = mk ?loc ?attrs (Pmty_typeof a)
   let extension ?loc ?attrs a = mk ?loc ?attrs (Pmty_extension a)
+
+  let strengthen mty lid =
+    let loc = mty.pmty_loc in
+    let mt =
+      { pmty_desc = Pmty_with (mty, [Pwith_module (lid,lid)]);
+        pmty_attributes = [];
+        pmty_loc = loc
+      }
+    in
+    let md =
+      { pmd_name = Location.mknoloc None;
+        pmd_type = mt;
+        pmd_attributes = [];
+        pmd_loc = loc
+      }
+    in
+    let payload = PSig [{psig_desc = Psig_module md; psig_loc = loc}] in
+    Pmty_extension (Location.mknoloc "strengthen", payload)
+
+  let unstrengthen = function
+    | {pmty_desc =
+        Pmty_extension (str, PSig [{psig_desc =
+          Psig_module {pmd_type = {pmty_desc =
+            Pmty_with (mty, [Pwith_module (_,lid)])}}}])}
+      when str.txt = "strengthen" -> Some (mty, lid)
+    | _ -> None
```

Comment:
Yes -- I'd be happy to throw that together.

Reply:
`functor (_ : [%extension.strengthen]) -> S with module L = L`

I'm not doubting you're right but I'd like to understand why this is better. I'm a bit worried that this would replace a signature with a functor type.

Reply:
It's a valid worry.

The problem with your original encoding is that ppxen don't generally look inside of extension nodes -- meaning your strengthened module types will be invisible to all ppx transformations, which could be bad.

Instead, the general-purpose extensions architecture is designed to be consulted essentially at every inspection of a parsetree type. The actual encoding I chose (in #141/#142) is `functor (_ : [%extensions.strengthen]) (_ : S) -> L`. (I decided not to use `with module L = L` to avoid duplication.) I know the `L` is a module, not a module type, but the parsetree just wants a `Longident.t` there, so it doesn't care about the distinction. (Though maybe this is bad, in case a ppx is looking for module identifiers? Feedback welcome, at https://github.com/ocaml-flambda/ocaml-jst/pull/142/files#r1131011021) 

As for your worry: we must be careful that every time we look at a module type, we check to see if it's the special form of a syntax extension. I've done this in #141. But the check that I've done it right is just "look to see if it looks right". Because we can't edit the parsetree, we can't really be sure of not making a mistake here. (Indeed, in #140, I found a few places where we might be getting this wrong for some expression extensions.)